state/protocol/protocol_state/epochs/happy_path_statemachine.go

```diff
@@ -124,25 +132,35 @@ func (u *HappyPathStateMachine) ProcessEpochSetup(epochSetup *flow.EpochSetup) (
 //     CAUTION: the HappyPathStateMachine is left with a potentially dysfunctional state when this error occurs. Do NOT call the Build method
 //     after such error and discard the HappyPathStateMachine!
 func (u *HappyPathStateMachine) ProcessEpochCommit(epochCommit *flow.EpochCommit) (bool, error) {
+	u.consumer.OnServiceEventReceived(epochCommit.ServiceEvent())
 	if u.state.NextEpoch == nil {
-		return false, protocol.NewInvalidServiceEventErrorf("protocol state has been setup yet")
+		err := protocol.NewInvalidServiceEventErrorf("protocol state has been setup yet")
+		u.consumer.OnInvalidServiceEvent(epochCommit.ServiceEvent(), err)
+		return false, err
 	}
 	if u.state.NextEpoch.CommitID != flow.ZeroID {
-		return false, protocol.NewInvalidServiceEventErrorf("protocol state has already a commit event")
+		err := protocol.NewInvalidServiceEventErrorf("protocol state has already a commit event")
+		u.consumer.OnInvalidServiceEvent(epochCommit.ServiceEvent(), err)
+		return false, err
 	}
 	err := protocol.IsValidExtendingEpochCommit(epochCommit, u.parentState.EpochProtocolStateEntry, u.parentState.NextEpochSetup)
 	if err != nil {
+		u.consumer.OnInvalidServiceEvent(epochCommit.ServiceEvent(), err)
```

Comment:
I was considering it but it's only possible for `Received` and maybe `Processed` event, because internal state machines might handle errors internally so we actually don't know what is a successful or an invalid event.